File: backend/app/auth/permissions.py
"""Role-permission map for the dashboard.

Endpoint guards check *permissions*, never role strings, so adding a new
role is a one-line edit to ROLE_PERMISSIONS. The role column in the
``users`` table also has a CHECK constraint listing known roles; if you
add a role here you also need to extend that constraint in a new
migration.
"""

from __future__ import annotations

from typing import Final

# --- permission constants ----------------------------------------------------
# Strings are stable identifiers — the frontend mirror in
# frontend/src/permissions.ts must use the exact same values.

DASHBOARD_PREVIEW: Final = "dashboard.preview"
DASHBOARD_VIEW: Final = "dashboard.view"
LIVE_REALTIME: Final = "live.realtime"
EVENT_LABEL_WRITE: Final = "event.label.write"
EVENT_DELETE: Final = "event.delete"
DEVICE_CONFIG_WRITE: Final = "device.config.write"
DEVICE_REGISTER: Final = "device.register"
USER_MANAGE: Final = "user.manage"

ALL_PERMISSIONS: Final[frozenset[str]] = frozenset(
    {
        DASHBOARD_PREVIEW,
        DASHBOARD_VIEW,
        LIVE_REALTIME,
        EVENT_LABEL_WRITE,
        EVENT_DELETE,
        DEVICE_CONFIG_WRITE,
        DEVICE_REGISTER,
        USER_MANAGE,
    }
)


# --- role → permission map ---------------------------------------------------
# Intentionally not cumulative across the whole hierarchy: guests see mock
# data (dashboard.preview), members see real data (dashboard.view) — those
# two don't bleed into each other. Higher tiers do inherit from the tier
# immediately below.

ROLE_PERMISSIONS: Final[dict[str, frozenset[str]]] = {
    "guest": frozenset({DASHBOARD_PREVIEW}),
    "member": frozenset({DASHBOARD_VIEW, LIVE_REALTIME}),
    "contributor": frozenset(
        {DASHBOARD_VIEW, LIVE_REALTIME, EVENT_LABEL_WRITE, EVENT_DELETE}
    ),
    "admin": frozenset(
        {
            DASHBOARD_VIEW,
            LIVE_REALTIME,
            EVENT_LABEL_WRITE,
            EVENT_DELETE,
            DEVICE_CONFIG_WRITE,
            DEVICE_REGISTER,
            USER_MANAGE,
        }
    ),
}

KNOWN_ROLES: Final[tuple[str, ...]] = tuple(ROLE_PERMISSIONS.keys())


def has_permission(role: str, permission: str) -> bool:
    return permission in ROLE_PERMISSIONS.get(role, frozenset())


def permissions_for(role: str) -> frozenset[str]:
    return ROLE_PERMISSIONS.get(role, frozenset())

File: frontend/src/permissions.ts
// Permission identifiers — must match backend/app/auth/permissions.py
// character-for-character. Endpoint guards on the backend check
// permissions, never role strings; UI gates do the same.

export const PERM = {
  DASHBOARD_PREVIEW: 'dashboard.preview',
  DASHBOARD_VIEW: 'dashboard.view',
  LIVE_REALTIME: 'live.realtime',
  EVENT_LABEL_WRITE: 'event.label.write',
  EVENT_DELETE: 'event.delete',
  DEVICE_CONFIG_WRITE: 'device.config.write',
  DEVICE_REGISTER: 'device.register',
  USER_MANAGE: 'user.manage',
  EVENT_CANDIDATE_MANAGE: 'event.candidate.manage',
} as const;

export type Permission = (typeof PERM)[keyof typeof PERM];

export type Role = 'guest' | 'member' | 'contributor' | 'admin';

// Mirror of ROLE_PERMISSIONS on the backend. Kept here for client-side
// gating only — the server is still the source of truth.
export const ROLE_PERMISSIONS: Record<Role, ReadonlySet<Permission>> = {
  guest: new Set<Permission>([PERM.DASHBOARD_PREVIEW]),
  member: new Set<Permission>([PERM.DASHBOARD_VIEW, PERM.LIVE_REALTIME]),
  contributor: new Set<Permission>([
    PERM.DASHBOARD_VIEW,
    PERM.LIVE_REALTIME,
    PERM.EVENT_LABEL_WRITE,
    PERM.EVENT_DELETE,
  ]),
  admin: new Set<Permission>([
    PERM.DASHBOARD_VIEW,
    PERM.LIVE_REALTIME,
    PERM.EVENT_LABEL_WRITE,
    PERM.EVENT_DELETE,
    PERM.DEVICE_CONFIG_WRITE,
    PERM.DEVICE_REGISTER,
    PERM.USER_MANAGE,
    PERM.EVENT_CANDIDATE_MANAGE,
  ]),
};

export function hasPermission(role: Role | null, perm: Permission): boolean {
  if (!role) return false;
  return ROLE_PERMISSIONS[role]?.has(perm) ?? false;
}
